src/rvsdg/from_egglog.rs

```diff
@@ -0,0 +1,231 @@
+        }
+    }
+
+    fn egglog_expr_to_rvsdg_ty(ty: &Expr) -> RvsdgType {
+        use Expr::*;
+        if let Call(func, args) = ty {
+            match (func.as_str(), &args.as_slice()) {
+                ("PrintState", []) => RvsdgType::PrintState,
+                ("Bril", [ty]) => RvsdgType::Bril(Self::egglog_expr_to_ty(ty)),
+                _ => panic!("expect an expression, got {ty}"),
+            }
+        } else {
+            panic!("expect an expression, got {ty}")
+        }
+    }
+
+    fn egglog_expr_to_literal(lit: &Expr) -> Literal {
+        use egglog::ast::{Expr::*, Literal::*};
+        if let Call(func, args) = lit {
+            match (func.as_str(), &args.as_slice()) {
+                ("Num", [Lit(Int(n))]) => Literal::Int(*n),
+                ("Float", [Lit(F64(n))]) => Literal::Float(f64::from(*n)),
+                ("Char", [Lit(String(s))]) => {
+                    assert_eq!(s.as_str().len(), 1);
+                    Literal::Char(s.as_str().chars().next().unwrap())
+                }
+                _ => panic!("expect a list, got {lit}"),
+            }
+        } else {
+            panic!("expect a list, got {lit}")
+        }
+    }
+
+    pub fn egglog_expr_to_function(expr: &Expr) -> RvsdgFunction {
+        use egglog::ast::{Expr::*, Literal::*};
+        if let Call(func, args) = expr {
+            match (func.as_str(), &args.as_slice()) {
+                ("Func", [Lit(String(name)), sig, Call(func_output, func_args)]) => {
+                    let args: Vec<RvsdgType> = vec_map(sig, Self::egglog_expr_to_rvsdg_ty);
+                    let n_args = args.len() - 1;
+
+                    let mut nodes = vec![];
+                    let (state, result) = match (func_output.as_str(), &func_args.as_slice()) {
+                        ("StateOnly", [state]) => {
+                            (Self::egglog_expr_to_operand(state, &mut nodes), None)
+                        }
+                        ("StateAndValue", [state, ty, result]) => {
+                            let state = Self::egglog_expr_to_operand(state, &mut nodes);
+                            let result = Self::egglog_expr_to_operand(result, &mut nodes);
+                            let ty = Self::egglog_expr_to_ty(ty);
+                            (state, Some((ty, result)))
+                        }
+                        _ => panic!("expect a function, got {expr}"),
+                    };
+                    RvsdgFunction {
+                        name: name.to_string(),
+                        n_args,
+                        args,
+                        nodes,
+                        result,
+                        state,
+                    }
+                }
+                _ => panic!("expect a function, got {expr}"),
+            }
+        } else {
+            panic!("expect a function, got {expr}")
+        }
+    }
+}
+
+fn vec_map<T>(inputs: &Expr, mut f: impl FnMut(&Expr) -> T) -> Vec<T> {
```

Comment:
I think this function should be documented. I don't think it's obvious from the signature the way in which it is a map operation.  I'm not positive about what it does, My guess is:

> Call `f` on each element of `inputs`, where `inputs` must be a fully expanded vector egglog vector expression.

Reply:
good idea, that's right